### PR TITLE
Add a ZkAddress-based constructor for ConfigAccessor

### LINKED ISSUE #591: Add a ZkAddress-based constructor to ConfigAccessor
In order to allow users to not take on the burden of creating an instance of HelixZkClient to use Helix data access APIs (like ConfigAccessor), we need to add a constructor that takes in a ZK connect string instead.

This approach has advantages such as 1) the user doesn't have to maintain the lifecycle of the ZkClient 2) add a layer of indirection between ZK and the user where Helix logic could be injected.

### PR DESCRIPTION
**Pull request by narendly**
### Issues

- [x] My PR addresses the following Helix issues and references them in the PR description:

Fixes #591 

### Description

- [x] Here are some details about my PR, including screenshots of any UI changes:

In order to allow users to not take on the burden of creating an instance of HelixZkClient to use Helix data access APIs (like ConfigAccessor), we need to add a constructor that takes in a ZK connect string instead.

This approach has advantages such as 1) the user doesn't have to maintain the lifecycle of the ZkClient 2) add a layer of indirection between ZK and the user where Helix logic could be injected.

Also close() is added so that the user could close after use.

### Tests

- [x] The following tests are written for this issue:

ConfigAccessor is covered by existing integration tests such as TestConfigAccessor.

- [x] The following is the result of the "mvn test" command on the appropriate module:

[INFO] Results:
[INFO] 
[ERROR] Failures: 
[ERROR]   TestCardDealingAdjustmentAlgorithmV2.testAlgorithmConstructor:128 expected:<9> but was:<6>
[ERROR] org.apache.helix.controller.strategy.crushMapping.TestCardDealingAdjustmentAlgorithmV2.testComputeMappingForDifferentReplicas(org.apache.helix.controller.strategy.crushMapping.TestCardDealingAdjustmentAlgorithmV2)
[ERROR]   Run 1: TestCardDealingAdjustmentAlgorithmV2.testComputeMappingForDifferentReplicas:274 Total movements: 4 != expected 8, replica: 1
[ERROR]   Run 2: TestCardDealingAdjustmentAlgorithmV2.testComputeMappingForDifferentReplicas:274 Total movements: 4 != expected 8, replica: 2
[ERROR]   Run 3: TestCardDealingAdjustmentAlgorithmV2.testComputeMappingForDifferentReplicas:274 Total movements: 14 != expected 21, replica: 3
[INFO]   Run 4: PASS
[INFO]   Run 5: PASS
[INFO] 
[ERROR]   TestAlertingRebalancerFailure.testTagSetIncorrect:174->checkResourceBestPossibleCalFailureState:310 expected:<true> but was:<false>
[INFO] 
[ERROR] Tests run: 882, Failures: 3, Errors: 0, Skipped: 1
[INFO] 
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  57:54 min
[INFO] Finished at: 2019-11-15T11:57:50-08:00

--------------

The tests pass when run individually.

### Commits

- [x] My commits all reference appropriate Apache Helix GitHub issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Code Quality

- [x] My diff has been formatted using helix-style.xml

### CODE REVIEW

File: helix-core/src/main/java/org/apache/helix/ConfigAccessor.java [hunk 1]
```diff
@@ -67,14 +68,30 @@
     // @formatter:on
   }
 
-  private final HelixZkClient zkClient;
+  private final HelixZkClient _zkClient;
+  // true if ConfigAccessor was instantiated with a HelixZkClient, false otherwise
+  // This is used for close() to determine how ConfigAccessor should close the underlying ZkClient
+  private boolean _usesExternalZkClient;
```
Comment (jiajunwang) [inline helix-core/src/main/java/org/apache/helix/ConfigAccessor.java]: final : )

File: helix-core/src/main/java/org/apache/helix/ConfigAccessor.java [hunk 2]
```diff
@@ -67,14 +68,30 @@
...
 
   /**
    * Initialize an accessor with a Zookeeper client
+   * Note: it is recommended to use the other constructor instead to avoid having to create a HelixZkClient.
    * @param zkClient
    */
   public ConfigAccessor(HelixZkClient zkClient) {
-    this.zkClient = zkClient;
+    this._zkClient = zkClient;
```
Comment (jiajunwang) [inline helix-core/src/main/java/org/apache/helix/ConfigAccessor.java]: This "this" is not necessary after we change the name to _zkClient.

Comment (narendly) [reply helix-core/src/main/java/org/apache/helix/ConfigAccessor.java]: Removed.

File: helix-core/src/main/java/org/apache/helix/ConfigAccessor.java (hunk 1 above)
Comment (narendly) [reply helix-core/src/main/java/org/apache/helix/ConfigAccessor.java]: Added

File: helix-core/src/main/java/org/apache/helix/ConfigAccessor.java [hunk 3]
```diff
@@ -67,14 +68,30 @@
...
+  // true if ConfigAccessor was instantiated with a HelixZkClient, false otherwise
+  // This is used for close() to determine how ConfigAccessor should close the underlying ZkClient
+  private final boolean _usesExternalZkClient;
 
   /**
    * Initialize an accessor with a Zookeeper client
+   * Note: it is recommended to use the other constructor instead to avoid having to create a HelixZkClient.
    * @param zkClient
    */
   public ConfigAccessor(HelixZkClient zkClient) {
```
Comment (lei-xia) [inline helix-core/src/main/java/org/apache/helix/ConfigAccessor.java]: Let us mark it  as @deprecated?

File: helix-core/src/main/java/org/apache/helix/ConfigAccessor.java [hunk 4]
```diff
@@ -786,16 +803,25 @@ private void updateInstanceConfig(String clusterName, String instanceName,
...
-      ZKUtil.createOrReplace(zkClient, zkPath, instanceConfig.getRecord(), true);
+      ZKUtil.createOrReplace(_zkClient, zkPath, instanceConfig.getRecord(), true);
     } else {
-      ZKUtil.createOrUpdate(zkClient, zkPath, instanceConfig.getRecord(), true, true);
+      ZKUtil.createOrUpdate(_zkClient, zkPath, instanceConfig.getRecord(), true, true);
+    }
+  }
+
+  /**
+   * Closes the ZkClient.
```
Comment (lei-xia) [inline helix-core/src/main/java/org/apache/helix/ConfigAccessor.java]: minor: close the ConfigAccessor to release the ZkClient or other stateful resources.

File: helix-core/src/main/java/org/apache/helix/ConfigAccessor.java (hunk 3 above)
Comment (narendly) [reply helix-core/src/main/java/org/apache/helix/ConfigAccessor.java]: Done

File: helix-core/src/main/java/org/apache/helix/ConfigAccessor.java (hunk 4 above)
Comment (narendly) [reply helix-core/src/main/java/org/apache/helix/ConfigAccessor.java]: Done

File: helix-core/src/main/java/org/apache/helix/ConfigAccessor.java (hunk 3 above)
Comment (narendly) [reply helix-core/src/main/java/org/apache/helix/ConfigAccessor.java]: Done

File: helix-core/src/main/java/org/apache/helix/ConfigAccessor.java [hunk 5]
```diff
@@ -67,14 +68,31 @@
     // @formatter:on
   }
 
-  private final HelixZkClient zkClient;
+  private final HelixZkClient _zkClient;
```
Comment (pkuwm) [inline helix-core/src/main/java/org/apache/helix/ConfigAccessor.java]: We are not expecting unrelated changes in a PR, right? I think if this name change is put in another PR, it would make this PR easier to review, because you can see the name change brings in noise(so many lines of change in this file).

Comment (narendly) [reply helix-core/src/main/java/org/apache/helix/ConfigAccessor.java]: What makes you think that this change is unrelated?
- This change is about ZkClient in ConfigAccessor.
- We have a new constructor that involves _zkClient. I didn't want to create a second private variable called _zkClient and leave zkClient as is for the sake of keeping the diff to a minimum. 

All in all, I think this diff is totally reasonable. Sure, you want to keep the change to a minimum. I appreciate the spirit.

**Comment by narendly** [comment]
This PR is ready to be merged, approved by @dasahcc .

Description: 

In order to allow users to not take on the burden of creating an instance of HelixZkClient to use Helix data access APIs (like ConfigAccessor), we need to add a constructor that takes in a ZK connect string instead.

This approach has advantages such as 1) the user doesn't have to maintain the lifecycle of the ZkClient 2) add a layer of indirection between ZK and the user where Helix logic could be injected.

Also close() is added so that the user could close after use.
